### DISCUSSION TITLE
Best practice for implementing a more robust Cholesky decomposition

### DISCUSSION THREAD

**Discussion by pipme** [jax-ml/jax#12356] [Q&A]
```python
import jax
import jax.numpy as jnp
from jax.scipy import linalg

@jax.jit
@jax.value_and_grad
def robust_choleksy(K):
    i = 0
    jitter = 1e-6
    L = linalg.cholesky(K)
    while jnp.any(jnp.isnan(L)):
        K += jitter*10**i
        L = linalg.cholesky(K)
        i += 1
    return np.sum(L)  # return a scalar for taking gradients
```
The code above doesn't work as expected. But how could I implement such a function properly, with `jit` and `value_and_grad`? Using `lax.while_loop` would throw `Reverse-mode differentiation does not work for lax.while_loop or lax.fori_loop. Try using lax.scan instead.` But I am not sure how could I achieve this with `lax.scan`. Thanks!

**Comment by hawkinsp** [comment]
My suggestion would be: write a custom gradient for this function?

You do not want to differentiate through the loop you wrote. You want to use the final output of the loop in the derivative. You can probably do this by writing a custom JVP rule.
